crates/red_knot_workspace/src/project/metadata.rs

```diff
@@ -0,0 +1,418 @@
+use ruff_db::system::{System, SystemPath, SystemPathBuf};
+use ruff_python_ast::name::Name;
+
+use crate::project::pyproject::{PyProject, PyProjectError};
+use crate::project::settings::Configuration;
+use red_knot_python_semantic::ProgramSettings;
+use thiserror::Error;
+
+#[derive(Debug, PartialEq, Eq)]
+#[cfg_attr(test, derive(serde::Serialize))]
+pub struct ProjectMetadata {
+    pub(super) name: Name,
+
+    pub(super) root: SystemPathBuf,
+
+    /// The resolved settings for this project.
+    pub(super) configuration: Configuration,
+}
+
+impl ProjectMetadata {
+    /// Creates a project with the given name and root that uses the default configuration options.
+    pub fn new(name: Name, root: SystemPathBuf) -> Self {
+        Self {
+            name,
+            root,
+            configuration: Configuration::default(),
+        }
+    }
+
+    /// Loads a project from a `pyproject.toml` file.
+    pub(crate) fn from_pyproject(
+        pyproject: PyProject,
+        root: SystemPathBuf,
+        base_configuration: Option<&Configuration>,
+    ) -> Self {
+        let name = pyproject.project.and_then(|project| project.name);
+        let name = name
+            .map(|name| Name::new(&*name))
+            .unwrap_or_else(|| Name::new(root.file_name().unwrap_or("root")));
+
+        // TODO: load configuration from pyrpoject.toml
+        let mut configuration = Configuration::default();
+
+        if let Some(base_configuration) = base_configuration {
+            configuration.extend(base_configuration.clone());
+        }
+
+        Self {
+            name,
+            root,
+            configuration,
+        }
+    }
+
+    /// Discovers the closest project at `path` and returns its metadata.
+    ///
+    /// The algorithm traverses upwards in the `path`'s ancestor chain and uses the following precedence
+    /// the resolve the project's root.
+    ///
+    /// 1. The closest `pyproject.toml` with a `tool.knot` section.
```

Comment:
I know this has come up in the CLI proposal. Skipping over projects without a `tool.knot` section is sort of important to support projects using uv's workspace layout where you have

```
root
| - packages
|   | -- a/
|   |     | - pyproject.toml
|   |     | - ...
|   | -- a/
|   |    | - pyproject.toml
|   |    | - ...
|  - src/ 
| - pyproject.toml (with `tool.knot` section)
```


In this case, it's important that the entire workspace can be checked as one. That's why running `knot` in `packages/a` should also pick up the configuration from the root, unless that one package overrides its own configuration.